### PR TITLE
When limiting VMs per template, count template VMs not all cloud VMs

### LINKED ISSUE #344: MaxDeploymentSize in templates is ignored and maxVirtualMachinesLimit is used as hard limit
### Jenkins and plugins versions report

<details>
  <summary>Environment</summary>

  
  ```text
Jenkins: 2.322
OS: Linux - 5.4.0-1062-azure
---
ace-editor:1.1
apache-httpcomponents-client-4-api:4.5.13-1.0
azure-credentials:198.vf9c2fdfde55c
azure-keyvault:131.v867845ef6ae9
azure-sdk:70.v63f6a95999a7
azure-vm-agents:799.va4c741108611
bootstrap4-api:4.6.0-3
bootstrap5-api:5.1.3-3
bouncycastle-api:2.25
branch-api:2.7.0
build-timestamp:1.0.3
build-user-vars-plugin:1.8
caffeine-api:2.9.2-29.v717aac953ff3
checks-api:1.7.2
cloud-stats:0.27
cloudbees-folder:6.16
command-launcher:1.6
configuration-as-code:1.54
credentials:2.6.2
credentials-binding:1.27
display-url-api:2.3.5
durable-task:493.v195aefbb0ff2
echarts-api:5.2.2-1
extended-read-permission:3.2
font-awesome-api:5.15.4-3
git:4.10.0
git-client:3.10.0
git-server:1.10
handlebars:3.0.8
jackson2-api:2.13.0-230.v59243c64b0a5
jaxb:2.3.0
jdk-tool:1.5
jquery-detached:1.2.1
jquery3-api:3.6.0-2
jsch:0.1.55.2
junit:1.53
ldap:2.7
lockable-resources:2.12
mailer:1.34
matrix-auth:2.6.8
matrix-project:1.19
momentjs:1.1.1
pipeline-build-step:2.15
pipeline-graph-analysis:1.12
pipeline-input-step:2.12
pipeline-milestone-step:1.3.2
pipeline-model-api:1.9.3
pipeline-model-declarative-agent:1.1.1
pipeline-model-definition:1.9.3
pipeline-model-extensions:1.9.3
pipeline-rest-api:2.19
pipeline-stage-step:2.5
pipeline-stage-tags-metadata:1.9.3
pipeline-stage-view:2.19
plain-credentials:1.7
plugin-util-api:2.5.1
popper-api:1.16.1-2
popper2-api:2.10.2-1
rebuild:1.32
resource-disposer:0.16
role-strategy:3.2.0
scm-api:2.6.5
script-security:1.78
snakeyaml-api:1.29.1
ssh-credentials:1.19
sshd:3.1.0
structs:1.24
throttle-concurrents:2.5
timestamper:1.15
trilead-api:1.0.13
workflow-aggregator:2.6
workflow-api:2.47
workflow-basic-steps:2.24
workflow-cps:2633.v6baeedc13805
workflow-cps-global-lib:548.v9085a486966a
workflow-durable-task-step:1101.vf832bc1ac745
workflow-job:2.42
workflow-multibranch:2.26
workflow-scm-step:2.13
workflow-step-api:2.24
workflow-support:3.8
ws-cleanup:0.39
  ```

</details>

### What Operating System are you using (both controller, and any agents involved in the problem)?

Jenkins in docker container

### Reproduction steps

I have the plugin configured as follows :
```
clouds:
  - azureVM:
      azureCredentialsId: "xxx"
      cloudName: "azure"
      cloudTags:
      - name: "deployer"
        value: "xxx"
      configurationStatus: "pass"
      deploymentTimeout: 1200
      existingResourceGroupName: "xxx"
      maxVirtualMachinesLimit: 90
      resourceGroupReferenceType: "existing"
      vmTemplates:
      - agentLaunchMethod: "SSH"
        agentWorkspace: "/var/jenkins"
        builtInImage: "Ubuntu 20.04 LTS"
        credentialsId: "jenkins"
        diskType: "managed"
        doNotUseMachineIfInitFails: true
        executeInitScriptAsRoot: true
        existingStorageAccountName: "xxx"
        imageReference:
          id: "xxx"
        imageTopLevelType: "advanced"
        javaPath: "java"
        labels: "xxx"
        location: "West Europe"
        maximumDeploymentSize: 10
        noOfParallelJobs: 1
        osType: "Linux"
        retentionStrategy:
          azureVMCloudRetentionStrategy:
            idleTerminationMinutes: 10
        storageAccountNameReferenceType: "existing"
        storageAccountType: "Standard_LRS"
        subnetName: "xxx"
        tags:
        - name: "application"
          value: "xxx"
        templateDesc: "xxx"
        templateName: "xxx"
        usageMode: EXCLUSIVE
        usePrivateIP: true
        virtualMachineSize: "Standard_E2s_v3"
        virtualNetworkName: "xxx"
        virtualNetworkResourceGroupName: "xxx"
```
As we can see the `maxVirtualMachinesLimit` is set to 90 and `maximumDeploymentSize` is set to 10. However `maximumDeploymentSize` seems to be ignored and up to 90 machines are spinned when using this configuration. Of course I have more images confgured this way and they all have `maximumDeploymentSize` set to 10. However this option is ignored.

In the jenkins FINE logs I can see : 
```
Mar 10, 2022 12:28:44 PM FINE com.microsoft.azure.vmagent.AzureVMCloud
Current estimated VM count: 90, quantity desired 2
Mar 10, 2022 12:28:44 PM INFO com.microsoft.azure.vmagent.AzureVMCloud provision
Not able to create 2 nodes, at or above maximum VM count of 90 and already 90 VM(s)

```
I checked the source code and there is a check which is never accessed and I cannot see it in the logs https://github.com/jenkinsci/azure-vm-agents-plugin/blob/a74208d4f7a1069427145e61b260372d8d6cd50c/src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java#L678

### Expected Results

`maximumDeploymentSize` is not ignored and correctly limits the virtual machine amount per template.

### Actual Results

`maximumDeploymentSize` is ignored and `maxVirtualMachinesLimit` is used as hard limit

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by pjdarton**
Fix #344

Implemented changes as per [my comment](https://github.com/jenkinsci/azure-vm-agents-plugin/issues/344#issuecomment-1158773947)

Added some basic unit-tests to verify the logic.

...but not tested it "for real" yet, so don't merge it yet!

### CODE REVIEW

File: src/main/java/com/microsoft/azure/vmagent/AzureVMCloudVerificationTask.java [hunk 1]
```diff
@@ -132,7 +133,7 @@ private static void verifyCloud(String cloudName) {
                         "AzureVMCloudVerificationTask: verify: cloud {0} already verified pass",
                         cloudName);
                 // Update the count.
-                cloud.setVirtualMachineCount(getVirtualMachineCount(cloud));
+                updateCloudVirtualMachineCounts(cloud);
```
Comment (pjdarton) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMCloudVerificationTask.java:136]: I'm not certain if this code is meant to be looking for/handling errors or not.
It looked like the old code would log errors and then simply return the previously-cached values making this a no-op in error situations.
I've made my new code return a boolean true on success and false on failure so _now_ this code could react to an error and do something about it if it wanted to.
...but I don't know if it should.

File: src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java [hunk 2]
```diff
@@ -1692,36 +1694,54 @@ public boolean isVMAliveOrHealthy(AzureVMAgent agent) throws AzureCloudException
...
+    }
+
+    /**
+     * Retrieves count of virtual machine in a azure subscription indexed by template name.
+     *
+     * @return {@link Map} of {@link String} template name to {@link Integer} count of VMs from that template.
+     * @throws ManagementException if we get an Azure error that isn't 404.
+     */
+    @Restricted(NoExternalUse.class)
+    Map<String, Integer> getVirtualMachineCountsByTemplate(final String cloudName, final String resourceGroupName) {
```
Comment (pjdarton) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java:1712]: I've not added unit-tests for this method.
It'll be exercised by the existing unit-tests for the original getVirtualMachineCount method, but there's nothing specific for this one at present.
...and I'd need a few hints as to how best to do that, as it'll need to "fake" a call to Azure (which the existing unit tests manage, but I don't fully understand how).

File: src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java [hunk 3]
```diff
@@ -1692,36 +1694,54 @@ public boolean isVMAliveOrHealthy(AzureVMAgent agent) throws AzureCloudException
...
-                        }
-                    } else {
-                        // keep backwards compatibility, omitting the resource created before updates
-                        // until all the resources has cloud tag
-                        count++;
-                    }
+                final String resourcesTag = tags.containsKey(Constants.AZURE_RESOURCES_TAG_NAME) ? tags.get(Constants.AZURE_RESOURCES_TAG_NAME) : null;
+                final String cloudTag = tags.containsKey(Constants.AZURE_CLOUD_TAG_NAME) ? tags.get(Constants.AZURE_CLOUD_TAG_NAME) : null;
+                final String templateTag = tags.containsKey(Constants.AZURE_TEMPLATE_TAG_NAME) ? tags.get(Constants.AZURE_TEMPLATE_TAG_NAME) : null;
+                if (resourcesTag == null || cloudTag == null || !deployTag.isFromSameInstance(new AzureUtil.DeploymentTag(resourcesTag))) {
```
Comment (pjdarton) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java:1722]: The previous code contained "kept for backwards compatibilty" stuff that I've removed.
The old code wasn't sure that everyone would have `cloudTag` set or not ... but that was in 2017, so I think we're safe now.

**Comment by timja** [comment]
> ...but not tested it "for real" yet, so don't merge it yet!

did you manage to test it?

**Comment by pjdarton** [comment]
No, not yet. Might be able to next week... but my ability to treat things "for real" is very limited as I don't have the ability to try it out on a "being used for real" Jenkins, only on a local only-used-by-me Jenkins which doesn't give the same degree of assurance as "real world" usage.

i.e. please don't solely rely on me for testing this.

(and, as I'm new to this plug-in code, feel free to point out where I've done things "differently" where you'd prefer me to "fit in" instead)

**Review by timja — APPROVED** [review]
I tested a few different scenarios and it all seemed to work fine.

I did manage to get above the max once but it was probably when I triggered a lot of jobs, given it's an approx count it should be fine

**Comment by pjdarton** [comment]
Awesome 👍 
That's got to be one of the quickest transitions from "I found a bug" to "fixed a bug" that I've encountered on Jenkins plugins (and that's even including ones that I maintain where I can drive the process end-to-end myself 😉).

So .... how soon before I can get my hands on an official release containing this fix? 😁
(I will see if I can get the interim build into our CI system as a temporary measure ... but we're using config-as-code which means that it's complicated to do anything other than point at official releases)

**Comment by pjdarton** [comment]
....and I've just seen the answer to my own question - you've already done one - so I'd guess that'll start showing up in the official Jenkins plugin updates later today 👍

**Comment by timja** [comment]
yup it's all automated :)

**Pull request merged**